### PR TITLE
Dataset definition

### PR DESCRIPTION
**Pull request by ZoeMZou**

### CODE REVIEW

File: analysis/create_project_actions.R
Comment (venexia) [inline analysis/create_project_actions.R:148]: It looks like we made a mistake with this previously as it is saying 15 actions when you have 5 actions. The following code should fix it:

```
# Return number of actions -----------------------------------------------------

count_run_elements <- function(x) {

  if (!is.list(x)) {
    return(0)
  }

  # Check if any names of this list are "run"
  current_count <- sum(names(x) == "run", na.rm = TRUE)
  
  # Recursively check all elements in the list
  return(current_count + sum(sapply(x, count_run_elements)))

}

print(paste0("YAML created with ",count_run_elements(actions_list)," actions."))
```

File: README.md
Comment (venexia) [inline README.md:14]: Specify which file as we will add more stuff to the lib directory as we go.

File: README.md
Comment (venexia) [inline README.md:1]: All the removed content from this file needs to be put back in.

File: README.md
Comment (venexia) [inline README.md:11]: Minor typo corrections below:

- If you are interested in how we defined our cohorts, we use the dataset definition script [dataset_definition_cohorts](analysis/dataset_definition_cohorts.py) to define a function that generates cohorts. This script imports all variables generated from [variables_cohorts](analysis/variables_cohorts.py) using the cohort index and end dates. This approach is used to generate three cohorts: pre-vaccination, vaccinated, and unvaccinated—found in [dataset_definition_prevax](analysis/dataset_definition_prevax.py), [dataset_definition_vax](analysis/dataset_definition_vax.py), and [dataset_definition_unvax](analysis/dataset_definition_unvax.py) respectively.

Comment (venexia) [reply README.md:11]: I am not sure what combined is referring to in this sentence: 'Extracted data is then combined to create our final cohorts, in the [preprocess data script](analysis/preprocess_data.R).'?

File: analysis/codelists.py
Comment (venexia) [inline analysis/codelists.py:142]: Comment formatting throughout the script could be made more consistent but looks good otherwise.

File: analysis/dataset_definition_cohorts.py
Comment (venexia) [inline analysis/dataset_definition_cohorts.py:1]: I think we need to change the name of this file to better describe what it does.

File: analysis/dataset_definition_dates.py
Comment (venexia) [inline analysis/dataset_definition_dates.py:1]: I think we need to change the name of this file to better describe what it does.

File: analysis/dataset_definition_prevax.py
Comment (venexia) [inline analysis/dataset_definition_prevax.py:17]: I think these need to be `end_date_exposure` and `end_date_outcome` as they are used throughout the rest of the code base like this. [This will apply to vax and unvax too.]

File: analysis/variables_cohorts.py
Comment (venexia) [inline analysis/variables_cohorts.py:186]: What does this comment mean by 'ONLY smoking codes'?

File: analysis/variables_cohorts.py
Comment (venexia) [inline analysis/variables_cohorts.py:179]: Why are these smoking codes under COVID-19 severity?

File: analysis/variables_cohorts.py
Comment (venexia) [inline analysis/variables_cohorts.py:214]: The outcomes are currently defined in primary care only as stated in the protocol. We usually also define outcomes in secondary care and death records - @rdenholm, do you know why the protocol says primary care only?

File: analysis/variables_cohorts.py
Comment (venexia) [inline analysis/variables_cohorts.py:250]: Existing respiratory conditions may also need to be defined in primary and secondary care, depending on the decision about outcomes.

File: analysis/variables_cohorts.py
Comment (venexia) [inline analysis/variables_cohorts.py:297]: It is okay that they fail the test - OpenSAFELY are aware of why we are using this variable and are happy for us to continue using it.

File: analysis/variables_cohorts.py
Comment (venexia) [inline analysis/variables_cohorts.py:447]: HbA1c is only required for diabetes specific project. Please remove comment.

File: analysis/variables_cohorts.py
Comment (venexia) [inline analysis/variables_cohorts.py:500]: I don't think all stroke is needed - I can't see it on the protocol and we usually use just ischaemic (expect for the cardiovascular specific project).

File: analysis/variables_cohorts.py
Comment (venexia) [inline analysis/variables_cohorts.py:518]: Can the temporary variables be defined by the code that combines them or do they need to be outside?

File: analysis/variables_cohorts.py
Comment (venexia) [inline analysis/variables_cohorts.py:538]: Do we need to check registration on index_date as well to confirm that they have the 6 months? I think this code tells you they were registered on the date six months before their index but not whether they remained registered.

**Review by venexia — CHANGES_REQUESTED** [review]
Hi @ZoeMZou - great work, well done! I have requested a couple of changes and there are some things to discuss at the next post-covid-events meeting. Also, please can you delete the file `dataset_definition.py` in this PR as it is no longer needed. Speak tomorrow.

File: analysis/active_analyses.R
Comment (ZoeMZou) [reply analysis/active_analyses.R:55]: Something like this? I will look into the core covars across post-covid-event projects
# Define core covariates (common across projects)
core_covars <- c(
    "cov_cat_ethnicity", "cov_cat_imd", "cov_num_consultation_rate", 
    "cov_bin_healthcare_worker", "cov_cat_smoking_status", "cov_bin_carehome_status"
)

# Define project-specific covariates (specific to the current project)
respiratory_covars <- c(
    "cov_bin_obesity", "cov_bin_history_pneumonia_snomed", "cov_bin_history_asthma_snomed", 
    "cov_bin_history_pulmonary_fibrosis_snomed", "cov_bin_ami", "cov_bin_all_stroke", 
    "cov_bin_dementia_combined", "cov_bin_liver_disease", "cov_bin_chronic_kidney_disease", 
    "cov_bin_cancer", "cov_bin_hypertension", "cov_bin_diabetes"
)

# Combine both vectors into a single vector for analysis
all_covars <- paste0(c(core_covars, project_covars), collapse = ";")

File: analysis/active_analyses.R
Comment (ZoeMZou) [reply analysis/active_analyses.R:7]: Done

File: analysis/active_analyses.R
Comment (ZoeMZou) [reply analysis/active_analyses.R:41]: Done

File: analysis/active_analyses.R
Comment (ZoeMZou) [reply analysis/active_analyses.R:500]: Done:
# Remove history of asthma from covariates for the sample with no preexisiting condition
df$covariate_other <- ifelse(
  df$population == "no_preexisting",
  sapply(
    strsplit(df$covariate_other, ";"), 
    function(covars) paste0(covars[covars != "cov_bin_history_asthma_snomed"], collapse = ";")
  ),
  df$covariate_other
)

File: README.md
Comment (ZoeMZou) [reply README.md:14]: Done: 
The [active_analyses](lib/active_analyses.rds) contains a list of active analyses.

File: README.md
Comment (ZoeMZou) [reply README.md:1]: Done

File: README.md
Comment (ZoeMZou) [reply README.md:11]: Directly copied from the mental health repo . Indeed nothing to combine for the current ehrQL repo. 
I will just say:
The extracted data is initially processed in the preprocess data script [preprocess data script](analysis/preprocess_data.R) to clean consultation rates, create a categorical variable for severe COVID-19, generate a flag variable for pre-existing respiratory conditions, and restrict columns, resulting in the final cohort.

File: analysis/codelists.py
Comment (ZoeMZou) [reply analysis/codelists.py:142]: Done

File: analysis/variables_dates.py
Comment (ZoeMZou) [reply analysis/variables_dates.py:55]: I extracted these reference dates from the study_dates dictionary in both scripts solely for calculation purposes, without importing them into the dataset, which is why the extraction codes appear identical. Is that correct?

File: analysis/variables_cohorts.py
Comment (ZoeMZou) [reply analysis/variables_cohorts.py:179]: We cannot use newly generated variables for calculations within the dictionary itself, so these variables need to be generated beforehand. While there isn’t strict logic required prior to defining the dictionary, incorporating these calculations within the dictionary structure is more reasonable and organized.

File: analysis/active_analyses.R
Comment (ZoeMZou) [reply analysis/active_analyses.R:64]: Done

File: analysis/dataset_definition_cohorts.py
Comment (ZoeMZou) [reply analysis/dataset_definition_cohorts.py:1]: According to our discussion, we will leave it for now.

File: analysis/dataset_definition_dates.py
Comment (ZoeMZou) [reply analysis/dataset_definition_dates.py:1]: According to our discussion, no need to change as it generated a dataset of dates as suggested in the name.

File: analysis/variables_cohorts.py
Comment (ZoeMZou) [reply analysis/variables_cohorts.py:186]: https://github.com/opensafely/early-inflammatory-arthritis/blob/069e61712fcc9a0c2ec2804ff36a9b773073291c/analysis/dataset_definition.py#L149-L160u

useful codes from @milanwiedemann

File: analysis/variables_dates.py
Comment (ZoeMZou) [reply analysis/variables_dates.py]: Done:
https://github.com/opensafely/post-covid-respiratory/blob/71046d8f9cec87faa7559a088620e6f765815bf5/analysis/variables_cohorts.py#L532-L551

File: analysis/variables_cohorts.py
Comment (ZoeMZou) [reply analysis/variables_cohorts.py]: Removed

File: analysis/variables_cohorts.py
Comment (ZoeMZou) [reply analysis/variables_cohorts.py]: Done:
https://github.com/opensafely/post-covid-respiratory/blob/ac685d23be5d0bafe24811cf9a952878dfae5f12/analysis/variables_cohorts.py#L520-L522

File: analysis/variables_cohorts.py
Comment (ZoeMZou) [reply analysis/variables_cohorts.py:501]: They need to be defined outside, so they can be included in the calculation here. Same as smoke variables. https://github.com/opensafely/post-covid-respiratory/pull/40#discussion_r1848038394

**Comment by ZoeMZou** [comment]
Hi @venexia ,
Thank you very much for your super helpful review! I have addressed all the points, except for the outcome variables, which are still pending a decision regarding the definition of outcomes in secondary care:

[Outcomes](https://github.com/opensafely/post-covid-respiratory/pull/40#discussion_r1838335669)
[Existing respiratory conditions](https://github.com/opensafely/post-covid-respiratory/pull/40#discussion_r1838337025)

Please could you have another review, thank you very much. 

Best wishes,
Zoe

File: analysis/active_analyses.R
Comment (venexia) [inline analysis/active_analyses.R:502]: This could be achieved using `outcomes_runall` and `outcomes_runmain` as for other projects if `preex_TRUE` was a subgroup rather than a population.

File: analysis/active_analyses.R
Comment (venexia) [inline analysis/active_analyses.R:508]: History of COPD covariate will need to be removed too.

File: analysis/variables_dates.py
Comment (venexia) [reply analysis/variables_dates.py]: We no longer need `tmp_dereg_date_before_index_date` as we have `has_follow_up_previous_6months`.

Also, `tmp_dereg_date_after_index_date` shouldn't be marked tmp as it is used for inclusion/exclusion. To that end, we should decide a prefix for this group of variable - perhaps 'inex_'?

File: analysis/variables_cohorts.py
Comment (venexia) [reply analysis/variables_cohorts.py]: Okay - do we know how this differs from what we were doing in terms of how many people change classification? I am conscious that we are trying to replicate at the moment so want to be making minimal number of changes.

File: analysis/variables_cohorts.py
Comment (venexia) [reply analysis/variables_cohorts.py:180]: I think the comments just need to be tidied up. You have:
1. Covid-19 prior to study start date (2020-01-01)
2. Covid-19 severity, including smoking codes

Please can you add a smoking heading?

File: analysis/variables_cohorts.py
Comment (venexia) [reply analysis/variables_cohorts.py:215]: Decision taken 03/12/2024 to include outcomes from primary care, secondary care and deaths as per other protocols. Recorded in issue https://github.com/opensafely/post-covid-respiratory/issues/45 and to be resolved in separate PR.

File: analysis/variables_cohorts.py
Comment (venexia) [reply analysis/variables_cohorts.py:251]: Decision taken 03/12/2024 to include outcomes from primary care, secondary care and deaths as per other protocols. Recorded in issue https://github.com/opensafely/post-covid-respiratory/issues/45 and to be resolved in separate PR.

File: README.md
Comment (venexia) [reply README.md]: Some futher edits below:

> -   If you are interested in how we defined our cohorts, we use the dataset definition script [dataset_definition_cohorts](analysis/dataset_definition_cohorts.py) to define a function that generates cohorts. This script imports all variables generated from [variables_cohorts](analysis/variables_cohorts.py) using the patient's index date, the cohort start date and the cohort end date. This approach is used to generate three cohorts: pre-vaccination, vaccinated, and unvaccinated—found in [dataset_definition_prevax](analysis/dataset_definition_prevax.py), [dataset_definition_vax](analysis/dataset_definition_vax.py), and [dataset_definition_unvax](analysis/dataset_definition_unvax.py), respectively. For each cohort, the extracted data is initially processed in the [preprocess data script](analysis/preprocess_data.R), which cleans consultation rate, creates a categorical variable for severe COVID-19, generates a flag variable for pre-existing respiratory conditions, and restricts the data to relevant variables.

File: analysis/active_analyses.R
Comment (venexia) [reply analysis/active_analyses.R]: I think we decided to go with the subgroup (rather than the population approach)...?

**Review by venexia — CHANGES_REQUESTED** [review]
Hi Zoe - this looks great, very almost there. I think the only necessary thing to resolve is to move from population > subgroup. Otherwise, I have added some minor fixes. Note: I made a commit to add an .Rproj to this repo so remember to pull before you make further changes!

File: analysis/active_analyses.R
Comment (ZoeMZou) [reply analysis/active_analyses.R]: Done

File: analysis/active_analyses.R
Comment (ZoeMZou) [reply analysis/active_analyses.R]: I have added "cov_bin_depression", "cov_bin_copd" to the protocol and this active_analyses script. In addition, I also defined the cov_bin_depression in variables_cohorts. I will leave the outcome related variables to be resolved in separate PR.

File: analysis/variables_dates.py
Comment (ZoeMZou) [reply analysis/variables_dates.py]: I suggest leaving the renaming for inclusion/exclusion variables to a separate PR, as I’ll need to adjust the preprocess.R script accordingly when processing this group of variables.

File: analysis/variables_cohorts.py
Comment (ZoeMZou) [reply analysis/variables_cohorts.py]: I think they are the equivalent codes to those used in previous Cohort Extractor projects:
https://github.com/opensafely/post-covid-mentalhealth/blob/ddb50bd8fc9e4e1a9855dc9f9a03991487c908aa/analysis/common_variables.py#L617-L630

File: analysis/variables_cohorts.py
Comment (ZoeMZou) [reply analysis/variables_cohorts.py:180]: I seee, got it! I revised the heading for smoking: 
    ## Define individual temporary variables (subgroup variables) first before using them in the dictrionary
    ## 1. Covid-19 prior to study start date (2020-01-01)
    ## 2. Covid-19 severity
    ## 3. Prepare for smoking status

File: analysis/active_analyses.R
Comment (ZoeMZou) [reply analysis/active_analyses.R:508]: In the seperate PR for considering outcomes from primary care, secondary care and deaths, I will also rename these covariates to `cov_bin_history_asthma` and `cov_bin_history_copd` .

File: analysis/active_analyses.R
Comment (ZoeMZou) [reply analysis/active_analyses.R]: As discussed, I have added the subgroup of preex_TRUE and preex_FALSE into loop and reflect that in the analysis coloum. So the above codes were deleted: 
https://github.com/opensafely/post-covid-respiratory/blob/ed683fbe43f338b0a9e37ad18f96623b4d43d216/analysis/active_analyses.R#L79-L109

**Comment by ZoeMZou** [comment]
Hi @venexia,

Thanks for the clarification on the comments earlier today. I’ve now completed the revisions for the PR. The main updates are:

1. active_analyses.R:

- Added `preex_FALSE` and `preex_TRUE` as subgroups in the loop, allowing the deletion of the population column and reflecting it within the analysis column.

2. variables_cohorts.py:

- Generate some variables from **preprocess.R** to **variables_cohorts.py** for consistency. 

- Renamed the inclusion/exclusion variables with the prefix inex_ for consistency.

Let me know whether they are okay to merge? 

Best wishes,
Zoe

File: analysis/variables_cohorts.py
Comment (venexia) [reply analysis/variables_cohorts.py]: Sorry, Zoe - I don't follow - if the code lists are identical, why did you swap them? And how many people change classification as a result of correcting the "ever smoke" category?

**Review by venexia — CHANGES_REQUESTED** [review]
Hi @ZoeMZou - some more renaming (sorry!) and some clarification needed re the changes to smoking. Things should get quicker from here - just trying to get things named in a way that will help us later.

File: analysis/variables_cohorts.py
Comment (ZoeMZou) [reply analysis/variables_cohorts.py]: Apologies for the confusion. The functionality remains identical, but the previous code required manually extracting the codelist for "ever smoke." In the updated version, I replaced it with the `filter_codes_by_category()` command, which I learned from Milan's code. This makes codes shorter.

File: analysis/active_analyses.R
Comment (ZoeMZou) [reply analysis/active_analyses.R:45]: Copied from the previous version. Can change here by extracting from JSON.

File: analysis/active_analyses.R
Comment (ZoeMZou) [reply analysis/active_analyses.R:496]: Great idea, thanks Venexia. 
Done:

File: analysis/variables_cohorts.py
Comment (ZoeMZou) [reply analysis/variables_cohorts.py:561]: Dont think so, I deleted the inex_died.

File: analysis/active_analyses.R
Comment (ZoeMZou) [reply analysis/active_analyses.R]: https://github.com/opensafely/post-covid-respiratory/blob/218c59f1a0169937cf63c226df534680563a6658/analysis/active_analyses.R#L67-L76

File: analysis/variables_cohorts.py
Comment (venexia) [reply analysis/variables_cohorts.py]: Sorry - my comment was unclear. `purpose` should be inex_ / cov_ / exp_ / out_ / qa_ / sub_ but I couldn't recall what you were using for the censoring conditions. As you have changed death date as well, I can see we hadn't come up with a prefix. Perhaps 'cens_'?

**Review by venexia — CHANGES_REQUESTED** [review]
Hi @ZoeMZou - great, two very small final things:
1. My comment was unclear about the use of the word 'purpose' in variable names - sorry about that. Please check the reply on the thread.
2. Please can we shorten `inex_bin_follow_up_previous_6months` to `inex_bin_6m_reg`?

**Comment by ZoeMZou** [comment]
Hi @venexia, no problem. Silly me, thought purpose_ means creating this on purpose for censoring. All revised for `cens_date_dereg` and `cens_date_death`. Renamed `inex_bin_6m_reg` as well.

**Review by venexia — APPROVED** [review]
Great job, Zoe. Approved - please merge when you are ready.

**Pull request merged**
